File: server/encrypted-object/recovery-service.ts
import type { EnvelopeEncryptionService } from '../crypto/envelope-service';
import { planStoredCiphertext } from './core';
import {
  completeVaultRecoveryDrill,
  decodeVaultRecoveryManifest,
  planVaultRecoveryDrill,
  type VaultRecoveryDrillReceipt,
  type VaultRecoveryScope,
} from './recovery-core';
import type { VaultRecoveryBackupPort } from './recovery-ports';
import { decodeEncryptedObjectCiphertext } from './service';

export type VaultRecoveryDrillResult =
  | { readonly kind: 'verified'; readonly receipt: VaultRecoveryDrillReceipt }
  | {
      readonly kind: 'blocked';
      readonly reason:
        | 'invalid-manifest'
        | 'backup-unavailable'
        | 'scope-mismatch'
        | 'invalid-timestamp'
        | 'retention-expired'
        | 'rotation-incomplete'
        | 'incomplete-checkpoint'
        | 'missing-object'
        | 'invalid-ciphertext'
        | 'authentication-failed'
        | 'invalid-evidence';
    };

export type VaultRecoveryDrillService = {
  run(input: {
    readonly scope: VaultRecoveryScope;
    readonly drilledAt: number;
  }): Promise<VaultRecoveryDrillResult>;
};

export function createVaultRecoveryDrillService(input: {
  readonly backup: VaultRecoveryBackupPort;
  readonly encryption: EnvelopeEncryptionService;
}): VaultRecoveryDrillService {
  return {
    async run(command) {
      let rawManifest: unknown;
      try {
        rawManifest = await input.backup.loadManifest(command.scope);
      } catch {
        return { kind: 'blocked', reason: 'backup-unavailable' };
      }
      let manifest;
      try {
        manifest = decodeVaultRecoveryManifest(rawManifest);
      } catch {
        return { kind: 'blocked', reason: 'invalid-manifest' };
      }
      const plan = planVaultRecoveryDrill({
        scope: command.scope,
        manifest,
        drilledAt: command.drilledAt,
      });
      if (plan.kind === 'blocked') return plan;

      const verifiedVersions = [];
      let verifiedObjects = 0;
      for (const metadata of manifest.objects) {
        let rawCiphertext: unknown;
        try {
          rawCiphertext = await input.backup.loadCiphertext({
            backupId: manifest.backupId,
            objectKey: metadata.objectKey,
          });
        } catch {
          return { kind: 'blocked', reason: 'backup-unavailable' };
        }
        if (rawCiphertext === undefined) {
          return { kind: 'blocked', reason: 'missing-object' };
        }
        if (!(rawCiphertext instanceof Uint8Array)) {
          return { kind: 'blocked', reason: 'invalid-ciphertext' };
        }
        let ciphertext;
        try {
          ciphertext = decodeEncryptedObjectCiphertext(rawCiphertext);
        } catch {
          return { kind: 'blocked', reason: 'invalid-ciphertext' };
        }
        const storedPlan = planStoredCiphertext({
          metadata,
          ciphertext,
          actualCiphertextBytes: rawCiphertext.byteLength,
        });
        if (storedPlan.kind === 'rejected') {
          return { kind: 'blocked', reason: 'invalid-ciphertext' };
        }
        let plaintext: Uint8Array;
        try {
          plaintext = await input.encryption.decrypt({
            keyring: manifest.keyring,
            context: {
              vaultId: manifest.vaultId,
              object: metadata.object,
              objectRevision: metadata.objectRevision,
            },
            ciphertext,
          });
        } catch {
          return { kind: 'blocked', reason: 'authentication-failed' };
        }
        const plaintextBytes = plaintext.byteLength;
        plaintext.fill(0);
        if (plaintextBytes !== metadata.plaintextBytes) {
          return { kind: 'blocked', reason: 'invalid-ciphertext' };
        }
        verifiedObjects += 1;
        verifiedVersions.push(metadata.dekVersion);
      }

      const completion = completeVaultRecoveryDrill({
        manifest,
        drilledAt: command.drilledAt,
        verifiedObjects,
        verifiedVersions,
      });
      return completion.kind === 'verified'
        ? completion
        : { kind: 'blocked', reason: 'invalid-evidence' };
    },
  };
}

File: backend/internal/encryptedobject/recovery_service.go
package encryptedobject

import (
	"context"

	"github.com/fukamu/notes/backend/internal/cryptocontent"
)

type RecoveryBackupPort interface {
	LoadManifest(context.Context, RecoveryScope) ([]byte, error)
	LoadCiphertext(context.Context, RecoveryBackupID, ObjectKey) ([]byte, bool, error)
}

type RecoveryDrillResultKind string

const (
	RecoveryResultVerified RecoveryDrillResultKind = "verified"
	RecoveryResultBlocked  RecoveryDrillResultKind = "blocked"
)

type RecoveryDrillResultReason string

const (
	RecoveryInvalidManifest       RecoveryDrillResultReason = "invalid-manifest"
	RecoveryBackupUnavailable     RecoveryDrillResultReason = "backup-unavailable"
	RecoveryResultScopeMismatch   RecoveryDrillResultReason = "scope-mismatch"
	RecoveryResultInvalidTime     RecoveryDrillResultReason = "invalid-timestamp"
	RecoveryResultExpired         RecoveryDrillResultReason = "retention-expired"
	RecoveryResultRotation        RecoveryDrillResultReason = "rotation-incomplete"
	RecoveryResultCheckpoint      RecoveryDrillResultReason = "incomplete-checkpoint"
	RecoveryMissingObject         RecoveryDrillResultReason = "missing-object"
	RecoveryInvalidCiphertext     RecoveryDrillResultReason = "invalid-ciphertext"
	RecoveryAuthenticationFailed  RecoveryDrillResultReason = "authentication-failed"
	RecoveryResultInvalidEvidence RecoveryDrillResultReason = "invalid-evidence"
)

type RecoveryDrillResult struct {
	Kind    RecoveryDrillResultKind   `json:"kind"`
	Reason  RecoveryDrillResultReason `json:"reason,omitempty"`
	Receipt *RecoveryDrillReceipt     `json:"receipt,omitempty"`
}

type RecoveryDrillService struct {
	backup     RecoveryBackupPort
	encryption EncryptionPort
}

func NewRecoveryDrillService(backup RecoveryBackupPort, encryption EncryptionPort) (*RecoveryDrillService, error) {
	if backup == nil || encryption == nil {
		return nil, ErrInvalidRecovery
	}
	return &RecoveryDrillService{backup: backup, encryption: encryption}, nil
}

func (service *RecoveryDrillService) Run(
	ctx context.Context,
	scope RecoveryScope,
	drilledAt int64,
) RecoveryDrillResult {
	if service == nil || service.backup == nil || service.encryption == nil || ValidateRecoveryScope(scope) != nil {
		return blockedRecoveryResult(RecoveryInvalidManifest)
	}
	manifestBytes, err := service.backup.LoadManifest(ctx, scope)
	if err != nil {
		return blockedRecoveryResult(RecoveryBackupUnavailable)
	}
	manifest, err := DecodeRecoveryManifest(manifestBytes)
	clear(manifestBytes)
	if err != nil {
		return blockedRecoveryResult(RecoveryInvalidManifest)
	}
	plan := PlanRecoveryDrill(scope, manifest, drilledAt)
	if plan.Kind == RecoveryDrillBlocked {
		return blockedRecoveryResult(resultReasonForPlan(plan.Reason))
	}
	verifiedVersions := make([]cryptocontent.DEKVersion, 0, len(manifest.Objects))
	verifiedObjects := int64(0)
	for _, metadata := range manifest.Objects {
		encoded, found, err := service.backup.LoadCiphertext(ctx, manifest.BackupID, metadata.ObjectKey)
		if err != nil {
			return blockedRecoveryResult(RecoveryBackupUnavailable)
		}
		if !found {
			return blockedRecoveryResult(RecoveryMissingObject)
		}
		actualBytes := int64(len(encoded))
		ciphertext, err := cryptocontent.DecodeEnvelopeCiphertext(encoded)
		clear(encoded)
		if err != nil || !StoredCiphertextMatches(
			metadata.CryptoVersion,
			metadata.DEKVersion,
			&metadata.CiphertextBytes,
			ciphertext,
			actualBytes,
		) {
			return blockedRecoveryResult(RecoveryInvalidCiphertext)
		}
		plaintext, err := service.encryption.Decrypt(ctx, manifest.Keyring, cryptocontent.ObjectContext{
			VaultID: manifest.VaultID, Kind: metadata.Object.Kind, ObjectID: metadata.Object.ObjectID,
			ObjectRevision: metadata.ObjectRevision,
		}, ciphertext)
		if err != nil {
			clear(plaintext)
			return blockedRecoveryResult(RecoveryAuthenticationFailed)
		}
		plaintextBytes := int64(len(plaintext))
		clear(plaintext)
		if plaintextBytes != metadata.PlaintextBytes {
			return blockedRecoveryResult(RecoveryInvalidCiphertext)
		}
		verifiedObjects++
		verifiedVersions = append(verifiedVersions, metadata.DEKVersion)
	}
	completion := CompleteRecoveryDrill(manifest, drilledAt, verifiedObjects, verifiedVersions)
	if completion.Kind != RecoveryDrillVerified || completion.Receipt == nil {
		return blockedRecoveryResult(RecoveryResultInvalidEvidence)
	}
	return RecoveryDrillResult{Kind: RecoveryResultVerified, Receipt: completion.Receipt}
}

func blockedRecoveryResult(reason RecoveryDrillResultReason) RecoveryDrillResult {
	return RecoveryDrillResult{Kind: RecoveryResultBlocked, Reason: reason}
}

func resultReasonForPlan(reason RecoveryDrillBlockReason) RecoveryDrillResultReason {
	switch reason {
	case RecoveryScopeMismatch:
		return RecoveryResultScopeMismatch
	case RecoveryInvalidTimestamp:
		return RecoveryResultInvalidTime
	case RecoveryRetentionExpired:
		return RecoveryResultExpired
	case RecoveryRotationIncomplete:
		return RecoveryResultRotation
	case RecoveryCheckpointIncomplete:
		return RecoveryResultCheckpoint
	default:
		return RecoveryResultInvalidEvidence
	}
}
